### ISSUE TITLE
backtest.py

### ISSUE THREAD

**Issue by crex-06** [crex-06/bist-backtest-1#1]
import math
import time
from pathlib import Path

import numpy as np
import pandas as pd
import yfinance as yf

# ============================================================
# BIST BACKTEST v1
# ============================================================
# Günlük OHLCV verisi kullanır.
#
# Strateji:
#   - EMA20 / EMA50 trend
#   - RSI momentum
#   - hacim anomalisi
#   - breakout
#   - ATR tabanlı stop
#
# Risk:
#   - İşlem başına portföyün %0.75'i
#   - Maksimum 5 pozisyon
#   - Maksimum %20 tek pozisyon
#   - Maksimum %70 yatırım
#   - %30 nakit
#
# ÖNEMLİ:
# KAP, emir defteri ve intraday order-flow henüz dahil değildir.
# Bu ilk sürüm günlük OHLCV ile ölçüm yapar.
# ============================================================

STARTING_CAPITAL = 50_000.0

RISK_PER_TRADE = 0.0075
MAX_POSITIONS = 5
MAX_POSITION_PCT = 0.20
MAX_INVESTED_PCT = 0.70

COMMISSION_RATE = 0.0010
SLIPPAGE_RATE = 0.0010

ATR_STOP_MULTIPLIER = 2.0

START_DATE = "2019-01-01"
END_DATE = None

# BIST'ten geniş bir likit hisse evreni.
# Yahoo Finance sembolleri .IS ile kullanılır.
SYMBOLS = [
    "AKBNK", "ALARK", "ARCLK", "ASELS", "ASTOR",
    "BIMAS", "BRSAN", "CCOLA", "DOAS", "DOHOL",
    "ECILC", "EKGYO", "ENKAI", "EREGL", "FENER",
    "FROTO", "GARAN", "GUBRF", "HALKB", "HEKTS",
    "ISCTR", "KCHOL", "KONTR", "KOZAA", "KOZAL",
    "KRDMD", "MGROS", "OYAKC", "PETKM", "PGSUS",
    "SAHOL", "SASA", "SISE", "SKBNK", "TAVHL",
    "TCELL", "THYAO", "TKFEN", "TOASO", "TUPRS",
    "ULKER", "VAKBN", "YKBNK"
]

def download_data(symbol):
    ticker = symbol + ".IS"

    try:
        df = yf.download(
            ticker,
            start=START_DATE,
            end=END_DATE,
            auto_adjust=False,
            progress=False,
            threads=False,
        )

        if df.empty:
            return None

        if isinstance(df.columns, pd.MultiIndex):
            df.columns = df.columns.get_level_values(0)

        needed = ["Open", "High", "Low", "Close", "Volume"]

        for col in needed:
            if col not in df.columns:
                return None

        df = df[needed].copy()
        df = df.dropna()

        return df

    except Exception as e:
        print(f"DATA ERROR {symbol}: {e}")
        return None

def indicators(df):

    df = df.copy()

    df["EMA20"] = df["Close"].ewm(span=20, adjust=False).mean()
    df["EMA50"] = df["Close"].ewm(span=50, adjust=False).mean()

    delta = df["Close"].diff()

    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)

    avg_gain = gain.rolling(14).mean()
    avg_loss = loss.rolling(14).mean()

    rs = avg_gain / avg_loss.replace(0, np.nan)
    df["RSI"] = 100 - (100 / (1 + rs))

    previous_close = df["Close"].shift(1)

    tr1 = df["High"] - df["Low"]
    tr2 = (df["High"] - previous_close).abs()
    tr3 = (df["Low"] - previous_close).abs()

    tr = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)

    df["ATR"] = tr.rolling(14).mean()

    df["VOL_MA20"] = df["Volume"].rolling(20).mean()

    df["HIGH20"] = df["High"].rolling(20).max().shift(1)

    df["RETURN20"] = df["Close"].pct_change(20)

    return df

def signal(row):

    if pd.isna(row["EMA20"]):
        return False

    trend = row["EMA20"] > row["EMA50"]

    momentum = 50 < row["RSI"] < 72

    volume = (
        row["Volume"] >
        row["VOL_MA20"] * 1.20
    )

    breakout = (
        not pd.isna(row["HIGH20"]) and
        row["Close"] > row["HIGH20"]
    )

    positive_return = row["RETURN20"] > 0.05

    score = sum([
        trend,
        momentum,
        volume,
        breakout,
        positive_return
    ])

    # En az 4/5 koşul
    return score >= 4

def run_backtest(data):

    cash = STARTING_CAPITAL

    positions = {}

    equity_curve = []

    trades = []

    dates = sorted(
        set(
            date
            for df in data.values()
            for date in df.index
        )
    )

    for date in dates:

        # ----------------------------------------------------
        # Önce mevcut pozisyonların değerini hesapla
        # ----------------------------------------------------

        portfolio_value = cash

        for symbol, pos in list(positions.items()):

            df = data[symbol]

            if date not in df.index:
                portfolio_value += pos["shares"] * pos["entry"]
                continue

            close = float(df.loc[date, "Close"])

            portfolio_value += pos["shares"] * close

        # ----------------------------------------------------
        # EXIT
        # ----------------------------------------------------

        for symbol in list(positions.keys()):

            pos = positions[symbol]
            df = data[symbol]

            if date not in df.index:
                continue

            row = df.loc[date]

            close = float(row["Close"])

            stop = pos["stop"]

            exit_reason = None

            if close <= stop:
                exit_reason = "STOP"

            elif close < float(row["EMA20"]):
                exit_reason = "TREND_BREAK"

            elif float(row["RSI"]) < 45:
                exit_reason = "MOMENTUM_LOSS"

            if exit_reason:

                exit_price = close * (1 - SLIPPAGE_RATE)

                gross = (
                    exit_price *
                    pos["shares"]
                )

                commission = gross * COMMISSION_RATE

                cash += gross - commission

                pnl = (
                    exit_price - pos["entry"]
                ) * pos["shares"]

                trades.append({
                    "Date": date,
                    "Symbol": symbol,
                    "Entry": pos["entry"],
                    "Exit": exit_price,
                    "Shares": pos["shares"],
                    "PnL": pnl - commission - pos["entry_commission"],
                    "Reason": exit_reason,
                })

                del positions[symbol]

        # ----------------------------------------------------
        # ENTRY
        # ----------------------------------------------------

        invested = 0.0

        for symbol, pos in positions.items():

            if date in data[symbol].index:

                close = float(
                    data[symbol].loc[date, "Close"]
                )

                invested += (
                    close *
                    pos["shares"]
                )

        invested_limit = portfolio_value * MAX_INVESTED_PCT

        available_slots = (
            MAX_POSITIONS -
            len(positions)
        )

        if available_slots > 0:

            candidates = []

            for symbol, df in data.items():

                if symbol in positions:
                    continue

                if date not in df.index:
                    continue

                row = df.loc[date]

                if signal(row):

                    candidates.append(
                        (
                            symbol,
                            float(row["Close"]),
                            float(row["ATR"]),
                            float(row["RSI"])
                        )
                    )

            # En güçlü RSI momentumundan başlayarak
            candidates.sort(
                key=lambda x: x[3],
                reverse=True
            )

            for symbol, close, atr, rsi in candidates:

                if len(positions) >= MAX_POSITIONS:
                    break

                if atr <= 0 or math.isnan(atr):
                    continue

                if invested >= invested_limit:
                    break

                stop_distance = atr * ATR_STOP_MULTIPLIER

                stop = close - stop_distance

                if stop <= 0:
                    continue

                risk_amount = (
                    portfolio_value *
                    RISK_PER_TRADE
                )

                shares = int(
                    risk_amount /
                    stop_distance
                )

                if shares <= 0:
                    continue

                entry_price = close * (
                    1 + SLIPPAGE_RATE
                )

                gross_cost = (
                    shares *
                    entry_price
                )

                commission = (
                    gross_cost *
                    COMMISSION_RATE
                )

                max_position_value = (
                    portfolio_value *
                    MAX_POSITION_PCT
                )

                if gross_cost > max_position_value:

                    shares = int(
                        max_position_value /
                        entry_price
                    )

                    gross_cost = (
                        shares *
                        entry_price
                    )

                    commission = (
                        gross_cost *
                        COMMISSION_RATE
                    )

                total_cost = (
                    gross_cost +
                    commission
                )

                if shares <= 0:
                    continue

                if cash < total_cost:
                    continue

                if invested + gross_cost > invested_limit:
                    continue

                cash -= total_cost

                invested += gross_cost

                positions[symbol] = {
                    "shares": shares,
                    "entry": entry_price,
                    "stop": stop,
                    "entry_commission": commission,
                    "entry_date": date,
                }

        # ----------------------------------------------------
        # END-OF-DAY EQUITY
        # ----------------------------------------------------

        equity = cash

        for symbol, pos in positions.items():

            df = data[symbol]

            if date in df.index:

                close = float(
                    df.loc[date, "Close"]
                )

                equity += (
                    pos["shares"] *
                    close
                )

        equity_curve.append({
            "Date": date,
            "Equity": equity,
            "Cash": cash,
            "Positions": len(positions),
        })

    return (
        pd.DataFrame(equity_curve),
        pd.DataFrame(trades)
    )

def statistics(equity, trades):

    if equity.empty:
        return {}

    equity["Return"] = equity["Equity"].pct_change()

    start = float(equity["Equity"].iloc[0])
    end = float(equity["Equity"].iloc[-1])

    total_return = end / start - 1

    days = (
        equity["Date"].iloc[-1] -
        equity["Date"].iloc[0]
    ).days

    years = max(days / 365.25, 1 / 365.25)

    cagr = (
        (end / start) ** (1 / years)
    ) - 1

    rolling_max = equity["Equity"].cummax()

    drawdown = (
        equity["Equity"] /
        rolling_max -
        1
    )

    max_drawdown = drawdown.min()

    daily_returns = equity["Return"].dropna()

    if daily_returns.std() > 0:

        sharpe = (
            daily_returns.mean() /
            daily_returns.std()
        ) * np.sqrt(252)

    else:
        sharpe = 0

    downside = daily_returns[
        daily_returns < 0
    ]

    if len(downside) > 0 and downside.std() > 0:

        sortino = (
            daily_returns.mean() /
            downside.std()
        ) * np.sqrt(252)

    else:
        sortino = 0

    if not trades.empty:

        wins = trades[
            trades["PnL"] > 0
        ]

        losses = trades[
            trades["PnL"] < 0
        ]

        win_rate = (
            len(wins) /
            len(trades)
        )

        gross_profit = wins["PnL"].sum()
        gross_loss = abs(
            losses["PnL"].sum()
        )

        profit_factor = (
            gross_profit /
            gross_loss
            if gross_loss > 0
            else np.inf
        )

    else:

        win_rate = 0
        profit_factor = 0

    avg_daily_pnl = daily_returns.mean() * end

    return {
        "Starting Capital": start,
        "Ending Capital": end,
        "Total Return": total_return,
        "CAGR": cagr,
        "Max Drawdown": max_drawdown,
        "Sharpe": sharpe,
        "Sortino": sortino,
        "Win Rate": win_rate,
        "Profit Factor": profit_factor,
        "Trades": len(trades),
        "Average Daily PnL TL": avg_daily_pnl,
        "Average Daily Return": daily_returns.mean(),
    }

def main():

    print("=" * 70)
    print("BIST BACKTEST v1")
    print("=" * 70)

    data = {}

    for i, symbol in enumerate(SYMBOLS, 1):

        print(
            f"[{i}/{len(SYMBOLS)}] Downloading {symbol}..."
        )

        df = download_data(symbol)

        if df is None or len(df) < 100:

            print(f"  SKIP {symbol}")
            continue

        df = indicators(df)

        data[symbol] = df

        time.sleep(0.2)

    print()
    print(
        f"Loaded {len(data)} / "
        f"{len(SYMBOLS)} symbols."
    )

    if not data:
        raise RuntimeError(
            "No market data could be downloaded."
        )

    equity, trades = run_backtest(data)

    stats = statistics(
        equity,
        trades
    )

    output = Path("results")
    output.mkdir(exist_ok=True)

    equity.to_csv(
        output / "equity_curve.csv",
        index=False
    )

    trades.to_csv(
        output / "trades.csv",
        index=False
    )

    pd.DataFrame([
        stats
    ]).to_csv(
        output / "summary.csv",
        index=False
    )

    print()
    print("=" * 70)
    print("RESULTS")
    print("=" * 70)

    for key, value in stats.items():

        if "Return" in key or "Rate" in key or "CAGR" in key or "Drawdown" in key:

            print(
                f"{key:25}: "
                f"{value * 100:.2f}%"
            )

        elif "PnL" in key:

            print(
                f"{key:25}: "
                f"{value:,.2f} TL"
            )

        else:

            print(
                f"{key:25}: "
                f"{value:,.4f}"
            )

    print()
    print(
        f"Average daily P&L: "
        f"{stats['Average Daily PnL TL']:,.2f} TL"
    )

    print()
    print("Results saved to /results")

if __name__ == "__main__":
    main()
